apps/src/labs/labRedux.ts

```diff
@@ -3,18 +3,108 @@
 // other existing redux store.  It also holds useful state for newer labs
 // that use LabContainer.
 
-import {createSlice, PayloadAction} from '@reduxjs/toolkit';
+import {
+  AnyAction,
+  createAsyncThunk,
+  createSlice,
+  PayloadAction,
+  ThunkDispatch,
+} from '@reduxjs/toolkit';
+import {Channel, ProjectManagerStorageType, Source} from './types';
+import LabRegistry from './LabRegistry';
+import ProjectManagerFactory from './projects/ProjectManagerFactory';
+import {
+  setProjectUpdatedAt,
+  setProjectUpdatedError,
+  setProjectUpdatedSaving,
+  setProjectUpdatedSaved,
+} from '../code-studio/projectRedux';
+import ProjectManager from './projects/ProjectManager';
 
 interface LabState {
+  // If we are currently loading a lab.
   isLoading: boolean;
   isPageError: boolean;
+  // channel for the current project, or undefined if there is no current project.
+  channel: Channel | undefined;
+  // last saved source for the current project, or undefined if we have not loaded or saved yet.
+  source: Source | undefined;
```

Comment:
Philosophical question: not all labs will have channels/sources (e.g. AI for Oceans, AI Lab).  Does it make sense for all of this logic to be in `labRedux` or is it worth considering some more specialized redux stores that labs can optionally use?

Reply:
Good question. For now I'm inclined to keep this here for simplicity's sake, but we could split this out if we find we have a large amount of data stored here.

Reply:
I also think it's a question of where we want the configurability to be. For example - I could see us having some sort of plugin/configuration system that's defined by the top level lab JS file and passed into ProjectContainer as a list of enabled/disabled features. Then if the "sources" feature is disabled, then we just skip fetching the source inside the thunk? Would be more complicated as with more complex or interdependent features, but just an idea.